sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -56,7 +57,23 @@ case class EnsureRequirements(
     // Ensure that the operator's children satisfy their output distribution requirements.
     var children = originalChildren.zip(requiredChildDistributions).map {
       case (child, distribution) if child.outputPartitioning.satisfies(distribution) =>
-        child
+        (child.outputPartitioning, distribution) match {
+          case (p: HashPartitioning, d: ClusteredDistribution) =>
+            if (conf.getConf(SQLConf.REQUIRE_ALL_CLUSTER_KEYS_FOR_SOLE_PARTITION) &&
+              requiredChildDistributions.size == 1 && !p.isPartitionedOnFullKeys(d)) {
+              // Add an extra shuffle for `ClusteredDistribution` even though its child
```

Comment:
The `if` branch actually means the requirement is not satisfied.

Either adding back `HashClusteredPartition` or adding a boolean into `ClusteredPartition` seems neater for me, because there're other existing call sites of `satisfies` and might be more future call sites such that we don't want to replicate special the logic there, e.g.,
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ValidateRequirements.scala#L34-L50

I'd prefer to adding `HashClusteredDistribution` back as it's conceptually a different requirement from `ClusteredDistribution`. Dynamic dispatching is more preferable than if-else for polymorphism, in general.

Reply:
Thanks, @c21!

It's probably discussed in the another comment thread.  But I'd like to reiterate that although currently it's a global config per a query, the examples I mentioned show that it's a physical plan alternative that shouldn't be ignored by default. 

The planner eventually probably wants to make a more automatic decision between HashClusteredDistribution and ClusteredDistribution at each operator level (at least, when stats are available). By that time, we will still need to bring `HashClusteredDistribution` back anyway, regardless how it is named and where the dispatch goes (a dedicated class or a bool field in ClusteredDistribution).

Reply:
For instance, if we have `HashClusteredDistribution`, we can put the following small heuristics here:
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowExecBase.scala#L48

```
if (partitionSpec.exists(**_.dataType.isInstanceOf[BooleanType]**)) {
      HashClusteredDistribution(partitionSpec) :: Nil
 } else {
      ClusteredDistribution(partitionSpec) :: Nil
 }
```

We'll still need `HashClusteredDistribution` anyway. Happy to learn what an alternative implementation for such a heuristic would look like.

Reply:
Can you explain what does this heuristic mean? why we make the exception when all partition key types are boolean?

Reply:
> when all partition key types are boolean?

The code means "when some partition key type is boolean, let's require partitioning with all keys".  
It's to avoid running a window operation and its sorting operation with parallelism<=3 (i.e., at most three partitions to run Window, one for `true`, one for `false`, one for `null`).

Reply:
> Happy to learn what an alternative implementation for such a heuristic would look like.

Alternatively I think this can be done as well, in a separate physical plan rule post `EnsureRequirements`, e.g. below is a sketch of rule:

```scala
object AddShuffleForBooleanWindow extends Rule[SparkPlan] {
  def apply(plan: SparkPlan): SparkPlan = {
    plan.transformDown {
      case w: WindowExecBase
        if w.partitionSpec.exists(_.dataType.isInstanceOf[BooleanType]) =>

        val distribution = w.requiredChildDistribution.head
        val clustering = distribution.asInstanceOf[ClusteredDistribution].clustering

        w.child.outputPartitioning match {
          case HashPartitioning(expressions, _)
            if isPartitionOnFullKeys(expressions, clustering) => w
          case p =>
            val newPartitioning = distribution.createPartitioning(p.numPartitions)
            w.withNewChildren(w.child match {
              case ShuffleExchangeExec(_, c, so) => Seq(ShuffleExchangeExec(newPartitioning, c, so))
              case _ => Seq(ShuffleExchangeExec(newPartitioning, w.child))
            })
        }
      case other => other
    }
  }

  private def isPartitionOnFullKeys(
      expressions: Seq[Expression], clustering: Seq[Expression]): Boolean = {
    expressions.length == clustering.length && expressions.zip(clustering).forall {
      case (l, r) => l.semanticEquals(r)
  }
}
```

Reply:
>> Alternatively I think this can be done as well, in a separate physical plan rule post EnsureRequirements

Does the rule create correctness issue? 
The rule changed the input shuffle of `w` without re-considering the downstream operators/shuffles of `w`, such that the distribution requirement of `w` may not be met with the newly injected ShuffleExchangeExec.

I think shuffle planning should better only be done in `EnsureRequirements`.

Reply:
> Does the rule create correctness issue?
The rule changed the input shuffle of w without re-considering the downstream operators/shuffles of w, such that the distribution requirement of a downstream operator of w may actually not be met with the newly injected ShuffleExchangeExec.

@sigmod - yes, this is mainly to show that we can do the same thing as physical plan rules. We can either put the logic inside `EnsureRequirements`, or run `EnsureRequirements` again inside the new rule (we did similar thing in `OptimizeSkewedJoin`). I was mainly wanted to point out there is alternative options to achieve same effect.

Reply:
> We can either put the logic inside EnsureRequirements, 
> or run EnsureRequirements again inside the new rule

Thanks for elaborating.  Ok, it's do-able, but feels more complex than necessary and maybe riskier than simply expressing a requirement. Since there has already been a requirement matching framework in EnsureRequirements, expressing a different requirement seems neater to me.

Reply:
```
if (partitionSpec.exists(_.dataType.isInstanceOf[BooleanType])) {
    HashClusteredDistribution(partitionSpec) :: Nil
 } else {
    ClusteredDistribution(partitionSpec) :: Nil
 }
```

I'm not sure if this is a good example, since you also need to check `child.outputPartitioning`, right? otherwise, the partition keys in the child output partitioning may not contain boolean at all but you'd still require a full partition key match, which may not be optimal. For instance:
```
WindowExec(partitionSpec=(x: bool, y: int, z: long))
  - child partitioned by y, z 
```

So in the end, it seems like this is not only expressing requirements but also doing some sort of matching similar to what's done in `EnsureRequirements`.

Reply:
> So in the end, it seems like this is not only expressing requirements but also doing some sort of matching

If we want to achieve such plans, I think we can (should?) still capture such things in requirements and `satisfies`, instead of baking complexities into `EnsureRequirements` or yet-another physical rule. E.g,

```
case class ClusteredDistribution(
    clustering: Seq[Expression],
    containing: Set[Expression],  /* satisfied partitioning must contain those expressions */
    disallowedCombinations: Set(Set[Expression]), /* disallow those combinations */
    requiredNumPartitions: Option[Int] = None) 
```

where the latter two specify what subset of possibilities are considered. There may be other ways to express the same thing. Then, @sunchao, the requirement to capture what you want looks like:
 ClusteredDistribution({x, y, z}, {}, {{x}}) 

If we have that, HashClusteredDistribution is just a special case and not needed, because 
HashClusteredDistribution({x, y, z}) = ClusteredDistribution({x, y, z},  {x, y, z}, {})

Without this augmented ClusteredDistribution,  HashClusteredDistribution is the only thing we can play with.  
That being said, we need more expressive power for ClusteredDistribution, either more fields or the previous HashClusteredDistribution :-)

Reply:
> If we want to achieve such plans, I think we can (should?) still capture such things in requirements and satisfies, instead of baking complexities into EnsureRequirements or yet-another physical rule. E.g,

Yes agreed. It's better to avoid case analysis in `EnsureRequirements` or some other rules.

On the augmented `ClusteredDistribution`, I'm not sure whether we'll need `containing` and `disallowedCombinations` for broader scenarios though, or it is sufficient to just consider simple knobs like `spark.sql.requireAllClusterKeysForHashPartition` for the majority use cases. But as you demonstrated, I think we can just keep `ClusteredDistribution`, evolve it and make it more expressive if there're strong requirements in future.